File: java/Ice/throughput/ThroughputI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import com.zeroc.demos.Ice.throughput.Demo.*;

public final class ThroughputI implements Throughput
{
    public ThroughputI()
    {
        _byteSeq = new byte[ByteSeqSize.value];

        _stringSeq = new String[StringSeqSize.value];
        for(int i = 0; i < StringSeqSize.value; ++i)
        {
            _stringSeq[i] = "hello";
        }

        _structSeq = new StringDouble[StringDoubleSeqSize.value];
        for(int i = 0; i < StringDoubleSeqSize.value; ++i)
        {
            _structSeq[i] = new StringDouble();
            _structSeq[i].s = "hello";
            _structSeq[i].d = 3.14;
        }

        _fixedSeq = new Fixed[FixedSeqSize.value];
        for(int i = 0; i < FixedSeqSize.value; ++i)
        {
            _fixedSeq[i] = new Fixed();
            _fixedSeq[i].i = 0;
            _fixedSeq[i].j = 0;
            _fixedSeq[i].d = 0;
        }
    }

    @Override
    public boolean needsWarmup(com.zeroc.Ice.Current current)
    {
        _warmup = false;
        return _needsWarmup;
    }

    @Override
    public void
    startWarmup(com.zeroc.Ice.Current current)
    {
        _warmup = true;
    }

    @Override
    public void endWarmup(com.zeroc.Ice.Current current)
    {
        _warmup = false;
        _needsWarmup = false;
    }

    @Override
    public void sendByteSeq(byte[] seq, com.zeroc.Ice.Current current)
    {
    }

    @Override
    public byte[] recvByteSeq(com.zeroc.Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupByteSeq;
        }
        else
        {
            return _byteSeq;
        }
    }

    @Override
    public byte[] echoByteSeq(byte[] seq, com.zeroc.Ice.Current current)
    {
        return seq;
    }

    @Override
    public void sendStringSeq(String[] seq, com.zeroc.Ice.Current current)
    {
    }

    @Override
    public String[] recvStringSeq(com.zeroc.Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupStringSeq;
        }
        else
        {
            return _stringSeq;
        }
    }

    @Override
    public String[] echoStringSeq(String[] seq, com.zeroc.Ice.Current current)
    {
        return seq;
    }

    @Override
    public void sendStructSeq(StringDouble[] seq, com.zeroc.Ice.Current current)
    {
    }

    @Override
    public StringDouble[] recvStructSeq(com.zeroc.Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupStructSeq;
        }
        else
        {
            return _structSeq;
        }
    }

    @Override
    public StringDouble[] echoStructSeq(StringDouble[] seq, com.zeroc.Ice.Current current)
    {
        return seq;
    }

    @Override
    public void sendFixedSeq(Fixed[] seq, com.zeroc.Ice.Current current)
    {
    }

    @Override
    public Fixed[] recvFixedSeq(com.zeroc.Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupFixedSeq;
        }
        else
        {
            return _fixedSeq;
        }
    }

    @Override
    public Fixed[] echoFixedSeq(Fixed[] seq, com.zeroc.Ice.Current current)
    {
        return seq;
    }

    @Override
    public void shutdown(com.zeroc.Ice.Current current)
    {
        current.adapter.getCommunicator().shutdown();
    }

    private byte[] _byteSeq;
    private String[] _stringSeq;
    private StringDouble[] _structSeq;
    private Fixed[] _fixedSeq;

    private byte[] _warmupByteSeq = new byte[1];
    private String[] _warmupStringSeq = new String[1];
    private StringDouble[] _warmupStructSeq = new StringDouble[] { new StringDouble() };
    private Fixed[] _warmupFixedSeq = new Fixed[] { new Fixed() };

    private boolean _needsWarmup = true;
    private boolean _warmup = false;
}

File: csharp/Ice/throughput/ThroughputI.cs
// Copyright (c) ZeroC, Inc.

using Demo;

internal sealed class ThroughputI : ThroughputDisp_
{
    public ThroughputI()
    {
        _byteSeq = new byte[ByteSeqSize.value];
        _stringSeq = new string[StringSeqSize.value];
        for (int i = 0; i < StringSeqSize.value; ++i)
        {
            _stringSeq[i] = "hello";
        }
        _structSeq = new StringDouble[StringDoubleSeqSize.value];
        for (int i = 0; i < StringDoubleSeqSize.value; ++i)
        {
            _structSeq[i] = new StringDouble();
            _structSeq[i].s = "hello";
            _structSeq[i].d = 3.14;
        }
        _fixedSeq = new Fixed[FixedSeqSize.value];
        for (int i = 0; i < FixedSeqSize.value; ++i)
        {
            _fixedSeq[i].i = 0;
            _fixedSeq[i].j = 0;
            _fixedSeq[i].d = 0;
        }
    }

    public override bool needsWarmup(Ice.Current current)
    {
        _warmup = false;
        return _needsWarmup;
    }

    public override void startWarmup(Ice.Current current)
    {
        _warmup = true;
    }

    public override void endWarmup(Ice.Current current)
    {
        _warmup = false;
        _needsWarmup = false;
    }

    public override void sendByteSeq(byte[] seq, Ice.Current current)
    {
    }

    public override byte[] recvByteSeq(Ice.Current current)
    {
        if (_warmup)
        {
            return _warmupByteSeq;
        }
        else
        {
            return _byteSeq;
        }
    }

    public override byte[] echoByteSeq(byte[] seq, Ice.Current current)
    {
        return seq;
    }

    public override void sendStringSeq(string[] seq, Ice.Current current)
    {
    }

    public override string[] recvStringSeq(Ice.Current current)
    {
        if (_warmup)
        {
            return _warmupStringSeq;
        }
        else
        {
            return _stringSeq;
        }
    }

    public override string[] echoStringSeq(string[] seq, Ice.Current current)
    {
        return seq;
    }

    public override void sendStructSeq(StringDouble[] seq, Ice.Current current)
    {
    }

    public override StringDouble[] recvStructSeq(Ice.Current current)
    {
        if (_warmup)
        {
            return _warmupStructSeq;
        }
        else
        {
            return _structSeq;
        }
    }

    public override StringDouble[] echoStructSeq(StringDouble[] seq, Ice.Current current)
    {
        return seq;
    }

    public override void sendFixedSeq(Fixed[] seq, Ice.Current current)
    {
    }

    public override Fixed[] recvFixedSeq(Ice.Current current)
    {
        if (_warmup)
        {
            return _warmupFixedSeq;
        }
        else
        {
            return _fixedSeq;
        }
    }

    public override Fixed[] echoFixedSeq(Fixed[] seq, Ice.Current current)
    {
        return seq;
    }

    public override void shutdown(Ice.Current current)
    {
        current.adapter.getCommunicator().shutdown();
    }

    private readonly byte[] _byteSeq;
    private readonly string[] _stringSeq;
    private readonly StringDouble[] _structSeq;
    private readonly Fixed[] _fixedSeq;

    private readonly byte[] _warmupByteSeq = new byte[1];
    private readonly string[] _warmupStringSeq = new string[1];
    private readonly StringDouble[] _warmupStructSeq = new StringDouble[] { new() };
    private readonly Fixed[] _warmupFixedSeq = new Fixed[1];

    private bool _needsWarmup = true;
    private bool _warmup = false;
}
